lib/ansible/cli/doc.py

```diff
@@ -71,7 +73,197 @@ class PluginNotFound(Exception):
+            path = to_text(b_path, errors='surrogate_or_strict')
+            collname = _get_collection_name_from_path(b_path)
+
+            roles_dir = os.path.join(path, 'roles')
+            if os.path.exists(roles_dir):
+                for entry in os.listdir(roles_dir):
+                    full_path = os.path.join(roles_dir, entry, 'meta', self.ROLE_ARGSPEC_FILE)
+                    if os.path.exists(full_path):
+                        if name_filters is None:
+                            found.add((entry, collname, path))
+                        else:
+                            # Name filters might contain a collection FQCN or not.
+                            for fqcn in name_filters:
+                                if len(fqcn.split('.')) == 3:
+                                    (ns, col, role) = fqcn.split('.')
+                                    if '.'.join([ns, col]) == collname and entry == role:
+                                        found.add((entry, collname, path))
+                                elif fqcn == entry:
+                                    found.add((entry, collname, path))
+        return found
+
+    def _create_role_list(self, roles_path, collection_filter=None):
+        """Return a dict describing the listing of all roles with arg specs.
+
+        :param role_paths: A tuple of one or more role paths.
+
+        :returns: A dict indexed by role name, with 'collection' and 'entry_points' keys per role.
+
+        Example return:
+
+            results = {
+               'roleA': {
+                  'collection': '',
+                  'entry_points': {
+                     'main': 'Short description for main'
+                  }
+               },
+               'a.b.c.roleB': {
+                  'collection': 'a.b.c',
+                  'entry_points': {
+                     'main': 'Short description for main',
+                     'alternate': 'Short description for alternate entry point'
+                  }
+               'x.y.z.roleB': {
+                  'collection': 'x.y.z',
+                  'entry_points': {
+                     'main': 'Short description for main',
+                  }
+               },
+            }
+
+        """
+        if not collection_filter:
+            roles = self._find_all_normal_roles(roles_path)
+        else:
+            roles = []
+        collroles = self._find_all_collection_roles(collection_filter=collection_filter)
+
+        result = {}
+
+        def build_summary(role, collection, argspec):
+            if collection:
+                fqcn = '.'.join([collection, role])
+            else:
+                fqcn = role
+            if fqcn not in result:
+                result[fqcn] = {}
+            summary = {}
+            summary['collection'] = collection
+            summary['entry_points'] = {}
+            for entry_point in argspec.keys():
+                entry_spec = argspec[entry_point] or {}
```

Comment:
Under what circumstances would the `or {}`? An empty entry point inside the `argument_specs.yml` file?

```yaml
main:
  short_description: My role does amazing things

other_entry:
```

Reply:
Correct. That's my attempt to keep the program from crashing if someone created an empty `argument_spec.yml` file (which is useless, but crashes aren't nice).